crates/stackable-telemetry/src/tracing.rs

```diff
@@ -0,0 +1,484 @@
+///
+/// ## Tracing exporter overrides
+///
+/// OTLP Exporter settings:
+///
+/// - `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`
+/// - `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT`
+/// - `OTEL_EXPORTER_OTLP_TRACES_COMPRESSION`
+/// - `OTEL_EXPORTER_OTLP_TRACES_HEADERS`
+///
+/// General Span and Trace settings:
+///
+/// - `OTEL_SPAN_ATTRIBUTE_COUNT_LIMIT`
+/// - `OTEL_SPAN_EVENT_COUNT_LIMIT`
+/// - `OTEL_SPAN_LINK_COUNT_LIMIT`
+/// - `OTEL_TRACES_SAMPLER` (Defaults to `parentbased_always_on`. If "traceidratio" or "parentbased_traceidratio", then `OTEL_TRACES_SAMPLER_ARG`)
+///
+/// Batch Span Processor settings:
+///
+/// - `OTEL_BSP_MAX_QUEUE_SIZE`
+/// - `OTEL_BSP_SCHEDULE_DELAY`
+/// - `OTEL_BSP_MAX_EXPORT_BATCH_SIZE`
+/// - `OTEL_BSP_EXPORT_TIMEOUT`
+/// - `OTEL_BSP_MAX_CONCURRENT_EXPORTS`
+///
+/// _See defaults in the opentelemetry_sdk crate under [trace::config][3] and [trace::span_processor][4]._
+///
+/// ## Log exporter overrides
+///
+/// OTLP exporter settings:
+///
+/// - `OTEL_EXPORTER_OTLP_LOGS_COMPRESSION`
+/// - `OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`
+/// - `OTEL_EXPORTER_OTLP_LOGS_TIMEOUT`
+/// - `OTEL_EXPORTER_OTLP_LOGS_HEADERS`
+///
+/// Batch Log Record Processor settings:
+///
+/// - `OTEL_BLRP_MAX_QUEUE_SIZE`
+/// - `OTEL_BLRP_SCHEDULE_DELAY`
+/// - `OTEL_BLRP_MAX_EXPORT_BATCH_SIZE`
+/// - `OTEL_BLRP_EXPORT_TIMEOUT`
+///
+/// _See defaults in the opentelemetry_sdk crate under [log::log_processor][5]._
+///
+/// [1]: tracing::Subscriber
+/// [2]: https://docs.rs/opentelemetry-otlp/latest/src/opentelemetry_otlp/exporter/mod.rs.html
+/// [3]: https://docs.rs/opentelemetry_sdk/latest/src/opentelemetry_sdk/trace/config.rs.html
+/// [4]: https://docs.rs/opentelemetry_sdk/latest/src/opentelemetry_sdk/trace/span_processor.rs.html
+/// [5]: https://docs.rs/opentelemetry_sdk/latest/src/opentelemetry_sdk/logs/log_processor.rs.html
+pub struct Tracing {
+    service_name: &'static str,
+    console_log_config: SubscriberConfig,
+    otlp_log_config: SubscriberConfig,
+    otlp_trace_config: SubscriberConfig,
+}
+
+impl Tracing {
+    pub fn builder() -> TracingBuilder<builder_state::PreServiceName> {
+        TracingBuilder::default()
+    }
+
+    /// Initialise the configured tracing subscribers, returning a guard that
+    /// will shutdown the subscribers when dropped.
+    pub fn init(self) -> Result<Tracing> {
+        let mut layers: Vec<Box<dyn Layer<Registry> + Sync + Send>> = Vec::new();
+
+        if self.console_log_config.enabled {
+            let env_filter_layer = EnvFilter::builder()
+                .with_default_directive(self.console_log_config.level_filter.into()) // TODO (@NickLarsenNZ): support Directives
+                .from_env_lossy();
+            let console_output_layer =
+                tracing_subscriber::fmt::layer().with_filter(env_filter_layer);
+            layers.push(console_output_layer.boxed());
+        }
+
+        if self.otlp_log_config.enabled {
+            // Convert log::Record to tracing::Event
```

Comment:
Yeah, so there is the `log` crate, which by default sets a NoOp logger.
Applications are then free to install a log implementation which _does something_ which the the log records emitted by the app and libraries.

Then there is the `tracing` crate, which does something similar and is supposed to supersede the `log` crate. Like `log`, applications are then free to install a subscriber implementation (eg: log to console with colours, or write to file, or ship via OTLP).
Its analogue to `log::Record` is `tracing::Event`, but there is also `tracing::Span` which adds other dimensions to _observability_ (that is: the time span, and hierarchy).

We are using `tracing` to instrument various parts of the code and emit events from our libraries and apps, however we might use external crates that have decided to use the `log` crate and so we need a way to handle that.

So we have two options for handling both `log::Record` and `tracing::Event` logs:
1. Setup two log bridges, one for `log` and one for `tracing` so we can ship them via OpenTelemetry, or
2. Convert `log::Record` to `tracing::Event`, and then only handle shipping `tracing::Event`.

The first option technically works, but only considers OpenTelemetry log exports, and not OpenTelemetry traces, or other `tracing` subscribers (like the console output subscriber).

I have now just seen a mistake where this is only enabled when log export is enabled, but we would like it installed regardless... I will fix that and post the commit hash here.

_Disclaimer: log/logging/tracing/events are very overloaded terms, so I apologise if it that has not helped make things clear_.

Reply:
This makes it a lot clearer - thank you for the explanation!

Reply:
Do we might want to add _some_ of that explanation to the comment @adwk67 linked to?

Reply:
I have removed tracing-log in db16e77867066f78c96213f65bc68c3ce84f32bb

As I was testing, I realised we were not seeing any `log` messages. After a bit of troubleshooting, I noticed that it only worked if it was initialised from the application and not from within the library. Maybe because instead of setting the `static mut LOGGER`, it is setting a boxed logger.

For any applications where we think we might need to capture `log::Record`, it will need to call `tracing_log::LogTracer::init()` directly.